### DISCUSSION TITLE
list_deployment_pipeline_stage_items Error

### DISCUSSION THREAD

**Discussion by Vijayinder** [microsoft/semantic-link-labs#380] [Q&A] [answered]
Hi Team, 

Has anyone tried to get list of items for deployment pipeline, Getting pipelines & Stages is working fine but when using Stage_items. I am getting below error. Actual API works with stage order but wrapper has been programmed to be name, I tried both name and order still getting error. 

API Call: list_deployment_pipeline_stage_items(deployment_pipeline="XYZ",stage_name="Production")
 KeyError                                  Traceback (most recent call last)
File ~/cluster-env/trident_env/lib/python3.10/site-packages/pandas/core/indexes/base.py:3653, in Index.get_loc(self, key)
   3652 try:
-> 3653     return self._engine.get_loc(casted_key)
   3654 except KeyError as err:

**Comment by m-kovalsky** [comment, marked as answer]
Thanks for raising the issue. Made a [fix](https://github.com/microsoft/semantic-link-labs/pull/381) for the next release.

**Reply by Vijayinder** [reply]
Thanks

**Comment by m-kovalsky** [comment]
This has been fixed. Try this in the latest release [0.9.0](https://github.com/microsoft/semantic-link-labs/releases/tag/0.9.0).

**Comment by Vijayinder** [comment]
Hello, I am still facing same issue, I have updated my version. Should it be stage name or number? I tried both without success. VijayOn 23 Jan 2025, at 1:02 am, m-kovalsky ***@***.***> wrote:﻿
This has been fixed. Try this in the latest release 0.9.0.

**Comment by m-kovalsky** [comment]
Please send the full error message in text

**Comment by Vijayinder** [comment]
Snippet showing that Version is updated 
![image](https://github.com/user-attachments/assets/abea5359-1b7a-4554-9795-d6e8efe0f382)

Worked out what I was doing wrong - Had to supply actual pipeline Name and also stage Name not ID's and number. I can now see the results but deployment datetime stamp is all coming up as Nat i.e. NULL

Output from this function
![image](https://github.com/user-attachments/assets/fb06a1a5-939e-41fd-87d4-8e82ae97a85f)

Output of Rest API
![image](https://github.com/user-attachments/assets/1bc6c0d9-a075-47e9-a6c4-93bd4b89e1d6)

**Comment by m-kovalsky** [comment]
Which API are you calling? You seem to be getting different properties than normal. [This](https://learn.microsoft.com/rest/api/fabric/core/deployment-pipelines/list-deployment-pipeline-stages?tabs=HTTP) is the API which is wrapped by this function.

**Comment by Vijayinder** [comment]
The one you mentioned is working I am looking at next step down calling funciton list_deployment_pipeline_stage_items which I believe is calling API GET [https://api.powerbi.com/v1.0/myorg/pipelines/{pipelineId}/stages/{stageOrder}/artifacts](https://learn.microsoft.com/en-us/rest/api/power-bi/pipelines/get-pipeline-stage-artifacts)

**Comment by m-kovalsky** [comment]
List_deployment_pipeline_stage_items is calling the api I sent in my previous message. It’s different from the one you are using.

**Comment by Vijayinder** [comment]
Looks like I am missing something. 

list_deployment_pipelines() wrapper for  https://learn.microsoft.com/en-au/rest/api/fabric/core/deployment-pipelines/list-deployment-pipelines?tabs=HTTP  - Lists all pipelines - This one is working great

list_deployment_pipeline_stages() wrapper for https://learn.microsoft.com/en-au/rest/api/fabric/core/deployment-pipelines/list-deployment-pipeline-stages?tabs=HTTP - This one list all stages in a pipeline - Which is working great - this is the one you shared earlier

list_deployment_pipeline_stage_items wrapper for https://learn.microsoft.com/en-au/rest/api/fabric/core/deployment-pipelines/list-deployment-pipeline-stage-items?tabs=HTTP - This one is expected to return all artifacts in the pipeline. The column headers returned by this are aligned to output of API only last deployment time is coming up as NULL

**Comment by m-kovalsky** [comment]
Please run the following code (filling in the 2 parameters) and show me the result for a deployment pipeline stage where the last deployment time shows as not null.

```python
import sempy.fabric as fabric

client = fabric.FabricRestClient()
deployment_pipeline_id = ''
stage_id = ''
response = client.get(f'/v1/deploymentPipelines/{deployment_pipeline_id}/stages/{stage_id}/items')
response.json()
```

**Comment by Vijayinder** [comment]
Looks like there is something going on with this particular API, [Microsoft example]([(https://learn.microsoft.com/en-au/rest/api/fabric/core/deployment-pipelines/list-deployment-pipeline-stage-items?tabs=HTTP#code-try-0)]) has last deployment datetime

![image](https://github.com/user-attachments/assets/7d854a8d-322a-4ccb-8dbc-40af70780b5f)

The response based on your code does not have lastdeployment datetime

![image](https://github.com/user-attachments/assets/b04ecb7f-5e66-4c89-989c-88349dbbe9a9)

Running Microsoft API directly is also not giving last deployment datetime now, so not sure if Msft are changing or updating this API
![image](https://github.com/user-attachments/assets/ee936fbc-adda-458b-8939-8e648f816108)

**Comment by m-kovalsky** [comment]
Yes that’s what I got too. It looks like it’s an issue with the api itself. It’s not showing the last deployment time. You should raise an issue with Microsoft support.

**Reply by Vijayinder** [reply]
Looks like Msft has changed the API back and now I can see last deployment time. Also noticed that it has changed back to stage order instead of Stage ID in the API call. GET https://api.powerbi.com/v1.0/myorg/pipelines/{pipelineId}/stages/{**stageOrder**}/artifacts

Running  code you provided is still the same, does not have last deployment time column and list_deployment_pipeline_stage_items is still returning NULL (NaT) for Last deployment Time

Snippet of API Call via Microsoft Rest API Page. 
![image](https://github.com/user-attachments/assets/a4b06f57-72ba-4efd-ab11-d9656f1efcd6)

**Comment by m-kovalsky** [comment]
You are comparing 2 different APIs. You are running a power bi api. The function in Semantic link labs is running the fabric api which I specified earlier in this thread.

**Comment by Vijayinder** [comment]
Thanks, confirming that Microsoft has advised they will fix fabric API in March.
